fms/datasets/__init__.py

```diff
@@ -123,3 +123,18 @@ def __iter__(self):
 
     def __len__(self):
         return len(self._map_ds)
+
+
+class PackedSequenceDataset(Dataset, SavableDataset):
+    def __init__(self, dataset: SavableDataset, max_seq_len: int):
+        self.dataset = dataset
+        self.max_seq_len = max_seq_len
+        self.buffer: List[Any] = []
+
+    def __iter__(self):
+        for example in self.dataset:
+            self.buffer.extend(example)
```

Comment:
hm... the instruction dataset adds its own, the arrow dataset doesn't.

Since the idea is to be composable, I think what we'll probably want to do is add another dataset that just adds bos/eos tokens to a wrapped dataset before passing to this one.

Reply:
Makes sense!